File: scripts/seed.py
"""Populate SQLite with demo org, site, berth, users, edge device (run from repo root)."""

from __future__ import annotations

import json
import secrets
import sys
from datetime import datetime, timezone
from decimal import Decimal
from pathlib import Path

import bcrypt

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "src"))

from maritime_fuel_tracker.db.base import Base  # noqa: E402
from sqlalchemy import select  # noqa: E402
from sqlalchemy.orm import Session  # noqa: E402

from maritime_fuel_tracker.db.models import (  # noqa: E402
    BerthLine,
    CalibrationRecord,
    Contract,
    Customer,
    Delivery,
    EdgeDevice,
    FuelGrade,
    FuelHub,
    HubSensor,
    Membership,
    MeterProfile,
    Organization,
    Site,
    Station,
    User,
    Vessel,
)
from maritime_fuel_tracker.db.session import get_engine, init_engine  # noqa: E402
from maritime_fuel_tracker.services.edge_auth import hash_api_key  # noqa: E402


def _now() -> datetime:
    return datetime.now(timezone.utc)


def _id() -> str:
    return secrets.token_hex(12)


def main() -> None:
    init_engine()
    eng = get_engine()
    Base.metadata.drop_all(bind=eng)
    Base.metadata.create_all(bind=eng)

    with Session(eng) as session:
        org = Organization(id=_id(), name="Demo Bunker Co", created_at=_now(), updated_at=_now())
        site = Site(
            id=_id(),
            organization_id=org.id,
            name="Pier 7 — High-flow berth",
            timezone="America/New_York",
            created_at=_now(),
            updated_at=_now(),
        )
        station = Station(
            id=_id(),
            site_id=site.id,
            code="STATION-1",
            display_name="Station 1 — Line A manifold",
            sort_order=1,
            created_at=_now(),
            updated_at=_now(),
        )
        berth = BerthLine(
            id=_id(),
            site_id=site.id,
            station_id=station.id,
            name="Line A",
            max_rate_m3h=3500.0,
            created_at=_now(),
            updated_at=_now(),
        )
        integration = {
            "protocol": "modbus_tcp",
            "host": "127.0.0.1",
            "port": 502,
            "holdingRegisters": {"massKg": 40001, "massRate": 40003},
            "notes": "Replace with vendor map.",
        }
        meter = MeterProfile(
            id=_id(),
            berth_line_id=berth.id,
            mode="CLAMP_ON_ULTRASONIC",
            integration_json=json.dumps(integration),
            created_at=_now(),
            updated_at=_now(),
        )
        session.add_all([org, site, station, berth, meter])

        admin_hash = bcrypt.hashpw(b"admin123", bcrypt.gensalt()).decode()
        it_hash = bcrypt.hashpw(b"itadmin123", bcrypt.gensalt()).decode()
        op_hash = bcrypt.hashpw(b"operator123", bcrypt.gensalt()).decode()
        fin_hash = bcrypt.hashpw(b"finance123", bcrypt.gensalt()).decode()

        admin = User(
            id=_id(),
            organization_id=org.id,
            email="admin@demo.local",
            password_hash=admin_hash,
            display_name="Org Admin",
            created_at=_now(),
            updated_at=_now(),
        )
        it = User(
            id=_id(),
            organization_id=org.id,
            email="itadmin@demo.local",
            password_hash=it_hash,
            display_name="IT / Control plane",
            created_at=_now(),
            updated_at=_now(),
        )
        op = User(
            id=_id(),
            organization_id=org.id,
            email="operator@demo.local",
            password_hash=op_hash,
            display_name="Berth Operator",
            created_at=_now(),
            updated_at=_now(),
        )
        fin = User(
            id=_id(),
            organization_id=org.id,
            email="finance@demo.local",
            password_hash=fin_hash,
            display_name="Finance",
            created_at=_now(),
            updated_at=_now(),
        )
        session.add_all([admin, it, op, fin])

        session.add_all(
            [
                Membership(id=_id(), user_id=admin.id, site_id=None, role="ORG_ADMIN", created_at=_now()),
                Membership(id=_id(), user_id=it.id, site_id=None, role="IT_ADMIN", created_at=_now()),
                Membership(id=_id(), user_id=op.id, site_id=site.id, role="OPERATOR", created_at=_now()),
                Membership(id=_id(), user_id=fin.id, site_id=site.id, role="FINANCE", created_at=_now()),
            ],
        )

        session.add_all(
            [
                FuelGrade(
                    id=_id(),
                    code="VLSFO",
                    description="Very Low Sulphur Fuel Oil",
                    sulfur_pct=0.5,
                    created_at=_now(),
                ),
                FuelGrade(
                    id=_id(),
                    code="MGO",
                    description="Marine Gas Oil",
                    sulfur_pct=0.1,
                    created_at=_now(),
                ),
            ],
        )
        session.flush()

        cust = Customer(
            id=_id(),
            organization_id=org.id,
            name="Oceanic Lines Ltd",
            billing_email="bunkers@oceanic.example",
            created_at=_now(),
            updated_at=_now(),
        )
        session.add(cust)
        session.add(
            Vessel(
                id=_id(),
                organization_id=org.id,
                name="MV Demo Trader",
                imo_number="9876543",
                flag="Panama",
                created_at=_now(),
                updated_at=_now(),
            ),
        )
        session.flush()

        session.add(
            Contract(
                id=_id(),
                organization_id=org.id,
                customer_id=cust.id,
                title="2026 VLSFO frame",
                currency="USD",
                basis="PER_MT",
                unit_price=Decimal("750.00"),
                tax_rate=Decimal("0.07"),
                rounding_kg=1000,
                rounding_m3=3,
                recon_variance_kg=Decimal("500"),
                effective_from=_now(),
                effective_to=None,
                created_at=_now(),
                updated_at=_now(),
            ),
        )
        session.flush()

        api_key = f"mbp_edge_{secrets.token_urlsafe(24)}"
        edge = EdgeDevice(
            id=_id(),
            site_id=site.id,
            name="Pier-7-edge-01",
            api_key_hash=hash_api_key(api_key),
            last_seen_at=None,
            created_at=_now(),
        )
        session.add(edge)
        session.flush()

        hub = FuelHub(
            id=_id(),
            site_id=site.id,
            station_id=station.id,
            berth_line_id=berth.id,
            code="PUMP-07",
            display_name="FuelTrace-class hub — Pier 7 Line A",
            firmware_version="4.2.1-demo",
            edge_device_id=edge.id,
            created_at=_now(),
            updated_at=_now(),
        )
        session.add(hub)
        session.flush()
        session.add_all(
            [
                HubSensor(
                    id=_id(),
                    hub_id=hub.id,
                    kind="FLOW_CLAMP_ON_ULTRASONIC",
                    label="Clamp-on ultrasonic pair",
                    channel_index=0,
                    install_json="{}",
                    is_active=True,
                    created_at=_now(),
                    updated_at=_now(),
                ),
                HubSensor(
                    id=_id(),
                    hub_id=hub.id,
                    kind="TEMP_PIPE_CLAMP",
                    label="Pipe temperature",
                    channel_index=1,
                    install_json="{}",
                    is_active=True,
                    created_at=_now(),
                    updated_at=_now(),
                ),
            ],
        )
        session.add(
            CalibrationRecord(
                id=_id(),
                meter_profile_id=meter.id,
                k_factor=Decimal("1.0000"),
                reference_note="Seed baseline K-factor.",
                created_at=_now(),
                created_by_id=None,
            ),
        )

        contract_row = session.execute(select(Contract).where(Contract.organization_id == org.id)).scalar_one()
        fg_row = session.execute(select(FuelGrade).where(FuelGrade.code == "VLSFO")).scalar_one()
        vessel_row = session.execute(select(Vessel).where(Vessel.organization_id == org.id)).scalar_one()
        session.add(
            Delivery(
                id=_id(),
                site_id=site.id,
                station_id=station.id,
                berth_line_id=berth.id,
                customer_id=cust.id,
                vessel_id=vessel_row.id,
                fuel_grade_id=fg_row.id,
                contract_id=contract_row.id,
                status="IN_PROGRESS",
                started_at=_now(),
                created_at=_now(),
                updated_at=_now(),
            ),
        )

        session.commit()

    print("\n=== Demo users (passwords) ===")
    print("admin@demo.local / admin123")
    print("itadmin@demo.local / itadmin123")
    print("operator@demo.local / operator123")
    print("finance@demo.local / finance123")
    print("\n=== Edge device API key ===")
    print(api_key)
    print("\nEdge ingest: POST http://localhost:8000/api/edge/v2/readings")
    with Session(eng) as s2:
        d = s2.execute(select(Delivery).limit(1)).scalar_one()
        print(f"\nDemo delivery id (use in edge JSON): {d.id}")
        print(f"Demo site id: {d.site_id}")
        print(f"Demo berth line id: {d.berth_line_id}")


if __name__ == "__main__":
    main()

File: _legacy/apps/web/prisma/seed.ts
import { HubSensorKind, PrismaClient, MeteringMode, PricingBasis, Role } from "@prisma/client";
import bcrypt from "bcryptjs";
import { createHash, randomBytes } from "crypto";
import { parseMeterIntegrationJson } from "../src/lib/schemas/meter-integration";

const prisma = new PrismaClient();

function hashApiKey(key: string) {
  return createHash("sha256").update(key, "utf8").digest("hex");
}

async function main() {
  await prisma.applicationErrorLog.deleteMany();
  await prisma.jobQueue.deleteMany();
  await prisma.stationMetricRollup.deleteMany();
  await prisma.auditEvent.deleteMany();
  await prisma.invoiceLine.deleteMany();
  await prisma.invoice.deleteMany();
  await prisma.reconciliationCase.deleteMany();
  await prisma.measurementReading.deleteMany();
  await prisma.edgeIdempotencyRecord.deleteMany();
  await prisma.bunkerDeliveryNote.deleteMany();
  await prisma.deliveryPreFlightItem.deleteMany();
  await prisma.delivery.deleteMany();
  await prisma.calibrationRecord.deleteMany();
  await prisma.hubSensor.deleteMany();
  await prisma.fuelHub.deleteMany();
  await prisma.edgeDevice.deleteMany();
  await prisma.meterProfile.deleteMany();
  await prisma.berthLine.deleteMany();
  await prisma.station.deleteMany();
  await prisma.membership.deleteMany();
  await prisma.user.deleteMany();
  await prisma.contract.deleteMany();
  await prisma.vessel.deleteMany();
  await prisma.customer.deleteMany();
  await prisma.site.deleteMany();
  await prisma.organization.deleteMany();
  await prisma.fuelGrade.deleteMany();

  const org = await prisma.organization.create({ data: { name: "Demo Bunker Co" } });
  const site = await prisma.site.create({
    data: { organizationId: org.id, name: "Pier 7 — High-flow berth", timezone: "America/New_York" },
  });
  const station = await prisma.station.create({
    data: {
      siteId: site.id,
      code: "STATION-1",
      displayName: "Station 1 — Line A manifold",
      sortOrder: 1,
    },
  });
  const berth = await prisma.berthLine.create({
    data: { siteId: site.id, stationId: station.id, name: "Line A", maxRateM3h: 3500 },
  });
  const integrationJson = JSON.stringify({
    protocol: "modbus_tcp",
    host: "127.0.0.1",
    port: 502,
    holdingRegisters: { massKg: 40001, massRate: 40003 },
    notes: "Replace with vendor map; Coriolis mass preferred for custody.",
  });
  parseMeterIntegrationJson(integrationJson);
  const meterProfile = await prisma.meterProfile.create({
    data: {
      berthLineId: berth.id,
      mode: MeteringMode.CLAMP_ON_ULTRASONIC,
      integrationJson,
    },
  });
  await prisma.calibrationRecord.create({
    data: {
      meterProfileId: meterProfile.id,
      kFactor: "1.0000",
      referenceNote: "Seed baseline K-factor (replace after in-situ calibration).",
    },
  });

  const adminHash = await bcrypt.hash("admin123", 10);
  const itHash = await bcrypt.hash("itadmin123", 10);
  const operatorHash = await bcrypt.hash("operator123", 10);
  const financeHash = await bcrypt.hash("finance123", 10);

  const admin = await prisma.user.create({
    data: {
      organizationId: org.id,
      email: "admin@demo.local",
      passwordHash: adminHash,
      displayName: "Org Admin",
    },
  });
  const itAdmin = await prisma.user.create({
    data: {
      organizationId: org.id,
      email: "itadmin@demo.local",
      passwordHash: itHash,
      displayName: "IT / Control plane",
    },
  });
  const operator = await prisma.user.create({
    data: {
      organizationId: org.id,
      email: "operator@demo.local",
      passwordHash: operatorHash,
      displayName: "Berth Operator",
    },
  });
  const finance = await prisma.user.create({
    data: {
      organizationId: org.id,
      email: "finance@demo.local",
      passwordHash: financeHash,
      displayName: "Finance",
    },
  });

  await prisma.membership.createMany({
    data: [
      { userId: admin.id, siteId: null, role: Role.ORG_ADMIN },
      { userId: itAdmin.id, siteId: null, role: Role.IT_ADMIN },
      { userId: operator.id, siteId: site.id, role: Role.OPERATOR },
      { userId: finance.id, siteId: site.id, role: Role.FINANCE },
    ],
  });

  await prisma.fuelGrade.create({
    data: { code: "VLSFO", description: "Very Low Sulphur Fuel Oil", sulfurPct: 0.5 },
  });
  await prisma.fuelGrade.create({
    data: { code: "MGO", description: "Marine Gas Oil", sulfurPct: 0.1 },
  });

  const customer = await prisma.customer.create({
    data: { organizationId: org.id, name: "Oceanic Lines Ltd", billingEmail: "bunkers@oceanic.example" },
  });
  await prisma.vessel.create({
    data: { organizationId: org.id, name: "MV Demo Trader", imoNumber: "9876543", flag: "Panama" },
  });

  await prisma.contract.create({
    data: {
      organizationId: org.id,
      customerId: customer.id,
      title: "2026 VLSFO frame",
      currency: "USD",
      basis: PricingBasis.PER_MT,
      unitPrice: "750.00",
      taxRate: "0.07",
      roundingKg: 1000,
      roundingM3: 3,
      reconVarianceKg: "500",
    },
  });

  const apiKey = `mbp_edge_${randomBytes(24).toString("base64url")}`;
  const edge = await prisma.edgeDevice.create({
    data: {
      siteId: site.id,
      name: "Pier-7-edge-01",
      apiKeyHash: hashApiKey(apiKey),
    },
  });
  await prisma.fuelHub.create({
    data: {
      siteId: site.id,
      stationId: station.id,
      berthLineId: berth.id,
      code: "PUMP-07",
      displayName: "FuelTrace-class hub — Pier 7 Line A",
      firmwareVersion: "4.2.1-demo",
      edgeDeviceId: edge.id,
      sensors: {
        create: [
          { kind: HubSensorKind.FLOW_CLAMP_ON_ULTRASONIC, label: "Clamp-on ultrasonic pair", channelIndex: 0 },
          { kind: HubSensorKind.TEMP_PIPE_CLAMP, label: "Pipe temperature", channelIndex: 1 },
          { kind: HubSensorKind.VIBRATION_PIPE, label: "Pump / cavitation accelerometer", channelIndex: 2 },
          { kind: HubSensorKind.TANK_LEVEL_ULTRASONIC, label: "Tank cross-check (roof mount)", channelIndex: 3 },
        ],
      },
    },
  });

  // eslint-disable-next-line no-console
  console.log("\n=== Demo users (passwords) ===");
  // eslint-disable-next-line no-console
  console.log("admin@demo.local / admin123");
  // eslint-disable-next-line no-console
  console.log("itadmin@demo.local / itadmin123  (control plane /admin)");
  // eslint-disable-next-line no-console
  console.log("operator@demo.local / operator123");
  // eslint-disable-next-line no-console
  console.log("finance@demo.local / finance123");
  // eslint-disable-next-line no-console
  console.log("\n=== Edge device API key (store in vault) ===");
  // eslint-disable-next-line no-console
  console.log(apiKey);
  // eslint-disable-next-line no-console
  console.log("\nPoint apps/edge-gateway at CLOUD_BASE_URL (this web app) and EDGE_API_KEY.");
  // eslint-disable-next-line no-console
  console.log("Edge v2 URL: /api/edge/v2/readings\n");
}

main()
  .then(async () => prisma.$disconnect())
  .catch(async (e) => {
    // eslint-disable-next-line no-console
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });
